src/main/java/io/hops/hopsworks/expat/migrations/featurestore/metadata/FeatureStoreMetadataMigration.java

```diff
@@ -0,0 +1,480 @@
+  }
+
+  private static String insertTagStmt(String artifactType) {
+    return INSERT_FS_TAGS_PART1 + artifactType + "_id" + INSERT_FS_TAGS_PART2;
+  }
+  
+  private String getKeywordsStmt(String artifactType) {
+    return GET_KEYWORDS_PART1 + artifactType + "_id" + GET_KEYWORDS_PART2;
+  }
+  
+  private static String insertKeywordStmt(String artifactType) {
+    return INSERT_FS_KEYWORDS_PART1 + artifactType + "_id" + INSERT_FS_KEYWORDS_PART2;
+  }
+  
+  private Map<String, String> readTags(byte[] tags) {
+    Map<String, String> tagsMap = new HashMap<>();
+    if (tags == null) {
+      return tagsMap;
+    }
+    String sTags = new String(tags, Charsets.UTF_8);
+    if(!Strings.isNullOrEmpty(sTags)) {
+      JSONArray tagsArr = new JSONArray(sTags);
+      for (int i = 0; i < tagsArr.length(); i++) {
+        JSONObject currentTag = tagsArr.getJSONObject(i);
+        if(currentTag.has("value")) {
+          tagsMap.put(currentTag.getString("key"), currentTag.getString("value"));
+        } else {
+          tagsMap.put(currentTag.getString("key"), "");
+        }
+      }
+    }
+    return tagsMap;
+  }
+  
+  private byte[] writeTags(Map<String, String> tags) {
+    List<JSONObject> tagsList = new ArrayList<>();
+    for(Map.Entry<String, String> entry : tags.entrySet()) {
+      JSONObject tag = new JSONObject();
+      tag.put("key", entry.getKey());
+      if(!Strings.isNullOrEmpty(entry.getValue())) {
+        tag.put("value", entry.getValue());
+      }
+      tagsList.add(tag);
+    }
+    return new JSONArray(tagsList).toString().getBytes(Charsets.UTF_8);
+  }
+  
+  private Set<String> readKeywords(byte[] keywords) {
+    Set<String> keywordsSet = new HashSet<>();
+    if (keywords == null) {
+      return keywordsSet;
+    }
+    String sKeywords = new String(keywords, Charsets.UTF_8);
+    if(!Strings.isNullOrEmpty(sKeywords)) {
+      JSONArray keywordsArr = new JSONArray(sKeywords);
+      for (int i = 0; i < keywordsArr.length(); i++) {
+        String currentKeyword = keywordsArr.getString(i);
+        keywordsSet.add(currentKeyword);
+      }
+    }
+    return keywordsSet;
+  }
+  
+  private byte[] writeKeywords(Set<String> keywords) {
+    return new JSONArray(keywords).toString().getBytes(Charsets.UTF_8);
+  }
+  
+  private String getArtifactPath(String artifactType, String projectName, String fsName, String name, Integer version){
+    switch(artifactType) {
+      case "feature_group": return getFGPath(fsName, name, version);
+      case "feature_view": return getFVPath(projectName, name, version);
+      case "training_dataset": return getTDPath(projectName, name, version);
+      default: throw new IllegalArgumentException("unknown artifact type:" + artifactType);
+    }
+  }
+  private String getFGPath(String fsName, String fgName, Integer fgVersion) {
+    return "/apps/hive/warehouse/"
+      + fsName + "_featurestore.db/"
+      + fgName + "_" + fgVersion;
```

Comment:
instead of doing `fgName + "_" + fgVersion`, `fvName + "_" + fvVersion`, `tdName + "_" + tdVersion` in each method could instead be done in `getArtifactPath` and provided as  a single variable

Reply:
extra method to save 3 lines of code doesn't seem worth it.
also it seems more readable to see the exact full path here. Having a method call makes you have to jump to the method to see the last dir